### PR TITLE
Add DG-subcell TciAndRollback action

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

The action checks if the solution at the future time is admissible, and if not,
undoes the time step and switches to the subcell solver.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/DgSubcell/Actions/TciAndRollback.hpp [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+namespace evolution::dg::subcell::Actions {
+/*!
+ * \brief Run the troubled-cell indicator on the candidate solution and perform
+ * the time step rollback if needed.
+ *
+ * The troubled-cell indicator (TCI) is given by the mutator `TciMutator` and
+ * can mutate tags in the DataBox, but should do so cautiously. The main reason
+ * that this is a mutator is because primitive variables, such as the pressure,
+ * are used to check if the solution is physical. In the relativistic case, even
+ * just whether or not the can be recovered is used as a condition. Note that
```
Comment (kidder) [inline src/Evolution/DgSubcell/Actions/TciAndRollback.hpp]: the can be -> the primitive variables can be

Comment (fmahebert) [inline src/Evolution/DgSubcell/Actions/TciAndRollback.hpp]: "whether or not the can be recovered" -> "whether or not the primitive variables can be recovered" ? or something else?

File: src/Evolution/DgSubcell/Actions/TciAndRollback.hpp [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+    using variables_tag = typename Metavariables::system::variables_tag;
+    using dt_variables_tag = db::add_tag_prefix<::Tags::dt, variables_tag>;
+
+    const ActiveGrid active_grid = db::get<Tags::ActiveGrid>(box);
+    ASSERT(active_grid == ActiveGrid::Dg,
+           "Must be using DG when calling TciAndRollback action.");
+
+    const Mesh<Dim>& dg_mesh = db::get<::domain::Tags::Mesh<Dim>>(box);
+    const auto& subcell_mesh = db::get<Tags::Mesh<Dim>>(box);
```
Comment (fmahebert) [inline src/Evolution/DgSubcell/Actions/TciAndRollback.hpp]: Is there a reason one variable is declared as `Mesh<Dim>` but the other as `auto` ?

File: src/Evolution/DgSubcell/Actions/TciAndRollback.hpp [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+          // Note: strictly speaking, to be conservative this should project uJ
+          // instead of u.
+          fd::project(inactive_vars_ptr, active_vars, dg_mesh,
+                      subcell_mesh.extents());
+        },
+        db::get<variables_tag>(box));
+
+    // Run RDMP TCI since no user info beyond the input file options are needed
+    // for that.
+    std::pair self_id{
```
Comment (fmahebert) [inline src/Evolution/DgSubcell/Actions/TciAndRollback.hpp:114]: Whoa, this is a cool C++17 feature I'd forgotten about

File: src/Evolution/DgSubcell/Actions/TciAndRollback.hpp [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+                 db::get<Tags::Inactive<variables_tag>>(box),
+                 self_neighbor_data.max_variables_values,
+                 self_neighbor_data.min_variables_values,
+                 subcell_options.rdmp_delta0(), subcell_options.rdmp_epsilon());
+
+    // If the RDMP TCI marked the candidate as acceptable, check with the
+    // user-specified TCI, since that could be stricter.
+    if (not cell_is_troubled) {
+      cell_is_troubled = db::mutate_apply<TciMutator>(
+          make_not_null(&box), subcell_options.persson_exponent());
```
Comment (fmahebert) [inline src/Evolution/DgSubcell/Actions/TciAndRollback.hpp:147]: For my understanding: most of the code looks like it's set up for generic `TciMutator`, but then here one of the args is related to the Persson TCI... is this a bug? or is the code not as generic as it looks?

File: src/Evolution/DgSubcell/Actions/TciAndRollback.hpp [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+                // Note: for strict conservation, we need to project uJ instead
+                // of just u.
+                std::get<0>(*initial_vars_ptr) =
+                    fd::project(std::get<0>(*initial_vars_ptr), dg_mesh,
+                                subcell_mesh.extents());
+              });
+        }
+      }
+      // We could/should project the _old_ primitive variables to the subcells
+      // to get a good initial guess.
```
Comment (fmahebert) [inline src/Evolution/DgSubcell/Actions/TciAndRollback.hpp]: It's not clear to me what this unattached comment is referring to.

File: tests/Unit/Evolution/DgSubcell/Actions/Test_TciAndRollback.cpp [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+        }
+      }
+    }
+  }
+}
+
+SPECTRE_TEST_CASE("Unit.Evolution.Subcell.Actions.TciAndRollback",
+                  "[Evolution][Unit]") {
+  // 1. Test RDMP passes/fails (check TciMutator not called on failure)
+  // 2. Test always_use_subcells (maybe?)
```
Comment (fmahebert) [inline tests/Unit/Evolution/DgSubcell/Actions/Test_TciAndRollback.cpp]: The added "(maybe?)" isn't conveying anything at all to me... can you rephrase?

File: tests/Unit/Evolution/DgSubcell/Actions/Test_TciAndRollback.cpp [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+  // 3. Test TciMutator passes/fails
+  //
+  // - history projected with correct size (one fewer)
+  // - active vars become projection of latest in history
+  // - active_grid is correct
+  // - did_rollback is correct
+  // - if self-start check initial value (and prims) were projected
```
Comment (fmahebert) [inline tests/Unit/Evolution/DgSubcell/Actions/Test_TciAndRollback.cpp:425]: I don't understand the relationship of the first numbered list (lines 414-16) and this list (418-422). Is one a sublist of the other? If not, why are there two lists? Some pre-list sentences describing what is being listed out would be very helpful here...

File: src/Evolution/DgSubcell/Actions/TciAndRollback.hpp [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+              tmpl::index_of<
+                  ActionList,
+                  ::Actions::Label<evolution::dg::subcell::Actions::Labels::
+                                       BeginSubcellAfterDgRollback>>::value +
+                  1};
+    }
+    // The unlimited DG solver has passed, so we can remove the current neighbor
+    // data.
+    db::mutate<subcell::Tags::NeighborDataForReconstructionAndRdmpTci<Dim>>(
+        make_not_null(&box), [](const auto nhbr_data_ptr) noexcept {
```
Comment (fmahebert) [inline src/Evolution/DgSubcell/Actions/TciAndRollback.hpp]: [style nitpick, feel free to ignore] I would slightly prefer 'nhbr' -> 'neighbor' throughout the PR (i.e. here and in the test file) because it's easier to read and not much longer

File: src/Evolution/DgSubcell/Actions/TciAndRollback.hpp (hunk 4 above)
Comment (nilsdeppe) [reply src/Evolution/DgSubcell/Actions/TciAndRollback.hpp:147]: Ah, good question! I'll add a comment, but the current reasoning is that you should use `exponent+1` for going back to DG, and rather than trying to get that correct via documentation, I figured it's easiest if I always pass the exponent in that people should be using based on what they specified in the input file

File: src/Evolution/DgSubcell/Actions/TciAndRollback.hpp (hunk 5 above)
Comment (nilsdeppe) [reply src/Evolution/DgSubcell/Actions/TciAndRollback.hpp]: I was on the fence about having this here, I'm just going to remove it. Basically, when switching from DG to subcell we need to do another primitive recovery. However, I think it's better to let the specific systems decide if or how that should be done. The thing this comment is trying to say is "If you need to do a primitive recovery on the subcells after the rollback, we can supply a high-order initial guess for the recovery scheme by projecting the primitives from the DG grid to the subcell grid"

**Comment by nilsdeppe** [comment]
I've rebased on develop and pushed a fixup. Thanks for the review :)

File: tests/Unit/Evolution/DgSubcell/Actions/Test_TciAndRollback.cpp [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+                                self_starting);
+        }
+      }
+    }
+  }
+}
+
+SPECTRE_TEST_CASE("Unit.Evolution.Subcell.Actions.TciAndRollback",
+                  "[Evolution][Unit]") {
+  // We to test the following cases:
```
Comment (fmahebert) [inline tests/Unit/Evolution/DgSubcell/Actions/Test_TciAndRollback.cpp]: "We to test" -> "We test" ?

**Review by fmahebert — COMMENTED** [review]
Fixups LGTM, please squash including this typo

**Comment by nilsdeppe** [comment]
I've rebased on develop, squashed in the fixup, and squashed in the wording change. Thanks for the review :)

**Review by fmahebert — APPROVED** [review]

**Review by kidder — APPROVED** [review]

File: src/Evolution/DgSubcell/Actions/TciAndRollback.hpp (hunk 1 above)
Comment (kidder) [reply src/Evolution/DgSubcell/Actions/TciAndRollback.hpp]: you already fixed this
